src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -302,60 +346,138 @@ val qualityFormat: Option[FastqQualityFormat] = None,
+private class SamRecordWriter(output: PathToBam,
+                              val header: SAMFileHeader,
+                              val umiTag: String) extends DemuxWriter {
+  private val writer = new SAMFileWriterFactory()
+    .setUseAsyncIo(DemuxFastqs.UseAsyncIo)
+    .setMaxRecordsInRam(DemuxFastqs.MaxRecordsInRamPerSamFileWriter)
+    .makeSAMOrBAMWriter(header, header.getSortOrder == SortOrder.unsorted, output.toFile)
+
+  private val rgId: String = this.header.getReadGroups.get(0).getId
+
+  def add(rec: DemuxRecord): Unit = {
+    val record  = new SAMRecord(header)
+    record.setReadName(rec.name)
+    record.setReadString(rec.bases)
+    record.setBaseQualities(rec.quals.getBytes)
+    record.setReadUnmappedFlag(true)
+    if (rec.pairedEnd) {
+      record.setReadPairedFlag(true)
+      record.setMateUnmappedFlag(true)
+      if (rec.readNumber == 1) record.setFirstOfPairFlag(true) else record.setSecondOfPairFlag(true)
+    }
+    record.setAttribute("BC", rec.sampleBarcode.orNull)
+    record.setAttribute(ReservedTagConstants.READ_GROUP_ID, rgId)
+    if (rec.molecularBarcode.nonEmpty) record.setAttribute(umiTag, rec.molecularBarcode.orNull)
 
-    val readGroup = new SAMReadGroupRecord(sample.sampleId)
-    readGroup.setSample(sample.sampleName)
-    readGroup.setLibrary(sample.libraryId)
-    readGroup.setPlatform("Illumina")
-    sample.description.foreach(readGroup.setDescription)
-    platformUnit.foreach(readGroup.setPlatformUnit)
-    sequencingCenter.foreach(readGroup.setSequencingCenter)
-    predictedInsertSize.foreach(readGroup.setPredictedMedianInsertSize)
-    runDate.foreach(readGroup.setRunDate)
-    platformModel.foreach(readGroup.setPlatformModel)
-
-    val header: SAMFileHeader = new SAMFileHeader
-    header.addReadGroup(readGroup)
-    header.setSortOrder(sortOrder)
-    comments.foreach(header.addComment)
-
-    val outputBam = if (isUnmatched) output.resolve(this.unmatched) else sampleOutputBam(output, sample)
-    val writer = new SAMFileWriterFactory()
-      .setUseAsyncIo(UseAsyncIo)
-      .setMaxRecordsInRam(MaxRecordsInRamPerSamFileWriter)
-      .makeSAMOrBAMWriter(header, header.getSortOrder == SortOrder.unsorted, outputBam.toFile)
-
-    SampleInfo(sample=sample, writer=writer, isUnmatched=isUnmatched)
+    writer.addAlignment(record)
   }
+
+  override def close(): Unit = writer.close()
 }
 
-/** A class to store information about a sample. */
-private[fastq] case class SampleInfo(sample: Sample, writer: SAMFileWriter, isUnmatched: Boolean = false) {
-  def header: SAMFileHeader = writer.getFileHeader
-  def rg: SAMReadGroupRecord = header.getReadGroups.get(0)
-  def rgId: String = rg.getId
+/** A writer that writes [[DemuxRecord]]s as [[FastqRecord]]s. */
+private class FastqRecordWriter(prefix: PathPrefix, val pairedEnd: Boolean) extends DemuxWriter {
+  private val writers: IndexedSeq[FastqWriter] = {
+    def f(ext: String): PathToFastq = PathUtil.pathTo(prefix + ext)
```

Comment:
I think this could be a little cleaner as:
```scala
  (if (pairedEnd) Seq("_R1.fastq.gz", "_R2.fastq.gz") else Seq(".fastq.gz"))
    .map(s => FastqWriter(Io.toWriter(PathUtil.pathTo(prefix + ext))))
    .toIndexedSeq
```

At the very least it could use our `Io.toWriter` instead of `IOUtil.openFileForBufferedWriting`